source/common/upstream/health_discovery_service.cc

```diff
@@ -301,18 +283,30 @@ HdsCluster::HdsCluster(Server::Admin& admin, Runtime::Loader& runtime,
       {admin, runtime_, cluster_, bind_config_, stats_, ssl_context_manager_, added_via_api_, cm,
        local_info, dispatcher, random, singleton_manager, tls, validation_visitor, api});
 
-  // iterate over every endpoint in every cluster.
+  // Temporary structure to hold Host pointers grouped by locality, to build
+  // initial_hosts_per_locality_.
+  std::vector<HostVector> hosts_by_locality;
+  // Iterate over every endpoint in every cluster.
   for (const auto& locality_endpoints : cluster_.load_assignment().endpoints()) {
+    // Add a locality grouping to the hosts sorted by locality.
+    hosts_by_locality.emplace_back();
```

Comment:
It might be helpful to have: 
hosts_by_locality.reserve(cluster_.load_assignment().endpoints_size());
hosts_by_locality.back().reserve(locality_endpoints.lb_endpoints_size());

Reply:
Thanks for the optimization, added.